lib/MirrorCache/WebAPI/Plugin/RenderFileFromMirror.pm

```diff
@@ -560,4 +578,106 @@ EOT
     return 1;
 }
 
+sub _render_btih() {
+    my ($c, $filename, $file) = @_;
+
+    unless($file->{md5}) {
+        $c->render(status => 404, text => "File not found");
+        return 1;
+    }
+
+    $c->render(text => "$filename " . _calc_btih($filename, $file));
+    return 1;
+}
+
+sub _render_magnet() {
+    my ($c, $url, $filename, $file) = @_;
+
+    unless($file->{piece_size}) {
+        $c->render(status => 404, text => "File not found");
+        return 1;
+    }
+
+    my $btih  = _calc_btih($filename, $file);
+    my $md5   = $file->{md5};
+    my $size  = $file->{size};
+    $filename = Mojo::Util::url_escape($filename);
+    $url      = Mojo::Util::url_escape($url);
+
+    $c->render(text => "magnet:?xt=urn:btih:$btih&amp;xt=urn:md5:$md5&amp;xl=$size&amp;dn=$filename&amp;as=$url");
+    return 1;
+}
+
+sub _render_torrent() {
+    my ($dm, $file, $mirrors_country, $mirrors_region, $mirrors_rest, $url) = @_;
+
+    my $c = $dm->c;
+
+    unless($file->{piece_size}) {
+        $c->render(status => 404, text => "File not found");
+        return 1;
+    }
+
+    my $tracker     = $ENV{MIRRORCACHE_TRACKER} // 'http://tracker.opensuse.org:6969/announce';
+    my $trackerlen  = length($tracker);
+    my $filename    = $file->{name};
+    my $filenamelen = length($filename);
+    my $size        = $file->{size};
+    my $mtime       = $file->{mtime};
+    my $md5         = $file->{md5};
+    my $piece_size  = $file->{piece_size};
+    my $sha1        = pack 'H*', $file->{sha1};
+    my $sha256      = pack 'H*', $file->{sha256};
+    my $pieces      = pack 'H*', $file->{pieces};
+
+    my $header = "d8:announce$trackerlen:$tracker" .
+                 "13:announce-listll$trackerlen:$tracker" .
+                 "ee7:comment$filenamelen:$filename" .
+                 "10:created by11:MirrorCache13:creation datei$mtime" .
+                 "e4:infod6:lengthi$size" .
+                 "e6:md5sum32:$md5" .
+                 "4:name$filenamelen:$filename" .
+                 "12:piece lengthi$piece_size" .
+                 "e6:pieces" . length($pieces) . ":";
```

Comment:
The leading `e` is probably the end-marker to the `i$integer` sequence in the line above.

maybe write as
```
"12:piece lengthi${piece_size}e"
```